tests/unit/lms/models/test_application_instance.py

```diff
@@ -33,6 +34,45 @@ def test_provisioning_is_not_nullable(self, db_session, application_instance):
 
         assert application_instance.provisioning is not None
 
+    def test_settings_defaults_to_an_empty_dict(self, db_session, application_instance):
+        settings = application_instance.settings
+
+        assert isinstance(settings, ApplicationSettings)
+        assert settings.data == {}
```

Comment:
I think `ApplicationSettings` should either have its own direct unit tests in an `_application_settings_test.py` file and the `ApplicationInstance` tests should patch `ApplicationSettings`, or `ApplicationSettings` should become an internal `_ApplicationSettings` helper within `application_instance.py` and then it can continue to be tested indirectly via `ApplicationInstance`.

Otherwise it's a bit strange, and breaks the usual layout pattern of our tests:

It's unclear where to expect to find the `ApplicationSettings` tests. It looks like it doesn't have any at first because there's no `_application_settings_test.py`, but in fact its tests can be found in another module's test file.

If `ApplicationSettings` might be re-used by another model (not just `ApplicationInstance`) in future, will the `ApplicationSettings` tests need to be repeated in the tests for the new model? If not, then it becomes even weirder that `ApplicationSettings` has two users and its tests can be found in the tests for one of its users but not the other.

If `ApplicationSettings` _isn't_ going to be reused then it should be an internal helper inside `application_instance.py`, then there's no problem with testing it indirectly via `ApplicationInstance`.

Reply:
I wanted to move this into a direction with application instance to make it explicit it's a private helper, or make it public but a few things stopped me in either direction:

 * I thought the PR would be rejected for being "too big" if I did move the application instance model
 * I think it actually could be used outside, for example a Canvas tool consumer object which is just passed the canvas portion for example...
 * ...but on the other hand I want to have deep tests that prove that manipulating the real settings object modifies the underlying SQL Alchemy. At this point if it was public we'd probably get stuck on enforced mocking, as I don't think we've resolved the "don't bother mocking PP model objects idea" yet. I thought the easiest way to avoid the conversation for longer would be to make it private for now.

So I'm kind of stuck with the current legal frame work.

My preferred solution would be:

 * Make the model public, test it in depth publicly
 * Define the unit of work to include the model: allow non mocking to prove changing the settings object modifies the underlying SQLAlchemy object